### PR TITLE
Add version control menu

### PR DESCRIPTION
**Pull request by sgugger**
This PR adds at the top of the navigation bar a menu to pick a version of the docs.

A few comments:

When switching version, the reader is sent on the same page of the docs in the older version (so it gives an error if the same page did not exist in this version of the docs). I don't know if this is preferable to the alternative (sending back to the index of the other version in the docs). Let me know what you think.

The menu will disappear once the reader goes to an older version of the docs (since it did not exist back then) unless we find a way to cherry pick in each release (but I doubt it's worth it).

Preview is [here](https://51918-155220641-gh.circle-artifacts.com/0/docs/_build/html/index.html). Side note, on a local build (like this one), the version appears as 'html', but it will be the right one once merged. Also, the preview/local build only has one version, so the links don't work there.

### CODE REVIEW

File: docs/source/_static/js/custom.js [hunk 1]
```diff
@@ -1,3 +1,26 @@
+// These two things need to be updated at each release for the version selector.
+// Last stable version
+let stableVersion = "v2.11.0"
+// Dictionary doc folder to label
+let versionMapping = {
```
Comment (LysandreJik) [inline docs/source/_static/js/custom.js:5]: We don't expect that to change, so it should be `const`

```suggestion
const stableVersion = "v2.11.0"
// Dictionary doc folder to label
const versionMapping = {
```

File: docs/source/_static/js/custom.js [hunk 2]
```diff
@@ -58,6 +81,63 @@ function addGithubButton() {
     document.querySelector(".wy-side-nav-search .icon-home").insertAdjacentHTML('afterend', div);
 }
 
+function addVersionControl() {
+    // To grab the version currently in view, we parse the url
+    let parts = location.toString().split('/');
```
Comment (LysandreJik) [inline docs/source/_static/js/custom.js:86]: This doesn't change either, best practice is to set it as a `const`
```suggestion
    const parts = location.toString().split('/');
```

File: docs/source/_static/js/custom.js [hunk 3]
```diff
@@ -58,6 +81,63 @@ function addGithubButton() {
...
+    var versionIndex = parts.length - 2
```
Comment (LysandreJik) [inline docs/source/_static/js/custom.js:87]: Should be `let`

```suggestion
    let versionIndex = parts.length - 2
```

File: docs/source/_static/js/custom.js [hunk 4]
```diff
@@ -58,6 +81,63 @@ function addGithubButton() {
...
+    // Main classes and models are nested so we need to go deeper
+    if (parts[versionIndex] == "main_classes" || parts[versionIndex] == "model_doc") {
+        versionIndex = parts.length - 3
+    } 
+    let version = parts[versionIndex];
```
Comment (LysandreJik) [inline docs/source/_static/js/custom.js:92]:
```suggestion
    const version = parts[versionIndex];
```

File: docs/source/_static/js/custom.js [hunk 5]
```diff
@@ -58,6 +81,63 @@ function addGithubButton() {
...
+
+    // Menu with all the links,
+    const versionMenu = document.createElement("div");
+
+    var htmlLines = [];
+    for (var key in versionMapping) {
```
Comment (LysandreJik) [inline docs/source/_static/js/custom.js:98]:
```suggestion
    for (const key in versionMapping) {
```

File: docs/source/_static/js/custom.js [hunk 6]
```diff
@@ -58,6 +81,63 @@ function addGithubButton() {
...
+    // Main classes and models are nested so we need to go deeper
+    if (parts[versionIndex] == "main_classes" || parts[versionIndex] == "model_doc") {
+        versionIndex = parts.length - 3
+    } 
+    let version = parts[versionIndex];
+
+    // Menu with all the links,
+    const versionMenu = document.createElement("div");
+
+    var htmlLines = [];
```
Comment (LysandreJik) [inline docs/source/_static/js/custom.js:97]: You can push to an array defined as a `const`

```suggestion
    const htmlLines = [];
```

**Review by LysandreJik — APPROVED** [review]
I'm in love with the result!

A few comments related to the best practices we have, essentially to have most variables set as constants.

File: docs/source/_static/js/custom.js [hunk 7]
```diff
@@ -58,6 +81,63 @@ function addGithubButton() {
...
+    versionButton.addEventListener("click", () => {
+        versionMenu.classList.toggle("version-show");
+    });
+
+    // Hide the menu when we click elsewhere
+    window.addEventListener("click", (event) => {
+        if (event.target != versionButton){
+            if (versionMenu.classList.contains('version-show')) {
+                versionMenu.classList.remove('version-show');
+            }
```
Comment (julien-c) [inline docs/source/_static/js/custom.js:124]:
```suggestion
            versionMenu.classList.remove('version-show');
```

File: docs/source/_static/js/custom.js [hunk 8]
```diff
@@ -58,6 +81,63 @@ function addGithubButton() {
...
+    if (parts[versionIndex] == "main_classes" || parts[versionIndex] == "model_doc") {
+        versionIndex = parts.length - 3
+    } 
+    const version = parts[versionIndex];
+
+    // Menu with all the links,
+    const versionMenu = document.createElement("div");
+
+    const htmlLines = [];
+    for (const key in versionMapping) {
```
Comment (julien-c) [inline docs/source/_static/js/custom.js:98]: I think in JS it's `for ... of`

`for ... in` is something slightly different

```suggestion
    for (const key of versionMapping) {
```

Comment (julien-c) [reply docs/source/_static/js/custom.js:98]: Ok no let me fix that comment :)

File: docs/source/_static/js/custom.js [hunk 9]
```diff
@@ -58,6 +81,63 @@ function addGithubButton() {
...
+        var value = versionMapping[key];
```
Comment (julien-c) [inline docs/source/_static/js/custom.js:99]:
```suggestion
    for (const [key, value] of Object.entries(versionMapping)) {
```

(yes it's slightly verbose)

**Comment by julien-c** [comment]
nice. In terms of UI you could also just have used a `<select>` element (maybe slightly more explicit UI) but I guess this works too

**Pull request merged**